### PR TITLE
Add mpi:// initialization method for torch.distributed

### LINKED ISSUE #8290: [feature request] MPI init_method for torch.distributed
That would be nice to have an init_method that is not socket-based, using MPI for initializing/side-channel communications. I am not talking about MPI backend (or cuda-aware MPI) here, but want to use mpi just for the init part. If we could use a custom `init_method` we could even use `mpi4py`.

## Code example

Ideally something like:

```python
            dist.init_process_group(
                backend="nccl",
                init_method='mpi://',
                world_size=self.world_size,
                rank=self.rank,
                group_name='mtorch'
            )
```

## System Info
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: Ubuntu 16.04.3 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 2.7
Is CUDA available: No
CUDA runtime version: 8.0.61
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.6.0.21
/usr/lib/x86_64-linux-gnu/libcudnn_static_v6.a

Versions of relevant libraries:
[pip] numpy (1.14.3)
[pip] numpydoc (0.8.0)
[pip] torch (0.4.0)
[pip] torchvision (0.2.1)
[conda] pytorch 0.4.0 py27_cuda8.0.61_cudnn7.1.2_1 pytorch
[conda] torchvision 0.2.1 py27_1 pytorch

### PR DESCRIPTION
**Pull request by pietern**
This initialization method can be used if a set of PyTorch processes
is launched through mpirun and MPI is available out of the box.
Currently, those runs must use the file://, env://, or tcp://
initialization methods, which can be cumbersome (e.g. reliance on a
shared filesystem, or having to know the IP address of rank 0).

If MPI support is available, and a set of processes is launched
through mpirun, the `init_method` keyword argument to
`torch.distributed.init_process_group` defaults to `mpi://`.

For those jobs, you can now simply call:

```
torch.distributed.init_process_group(backend='nccl')
```

The test suite was adapted to support testing of the Gloo and NCCL
backends when using the MPI initialization method.

Closes #8290.

### CODE REVIEW

File: torch/distributed/_object.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    buf = io.BytesIO()
+    torch.save(obj, buf)
+    return torch.ByteTensor(list(buf.getvalue()))
+
+
+def _tensor_to_object(tensor):
+    """
+    Deserialize Python object from torch.ByteTensor.
+    """
+    buf = io.BytesIO(bytearray(tensor.tolist()))
```
Comment (apaszke) [inline torch/distributed/_object.py]: Ugh, `tolist` is likely to be awfully slow. Can't we e.g. convert to numpy (if available), and use the buffer interface (e.g. `memoryview`) to initialize the IO buffer?

File: torch/distributed/_object.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import torch
+
+
+def _object_to_tensor(obj):
+    """
+    Serialize Python object to torch.ByteTensor.
+    """
+    buf = io.BytesIO()
+    torch.save(obj, buf)
+    return torch.ByteTensor(list(buf.getvalue()))
```
Comment (apaszke) [inline torch/distributed/_object.py]: Might be worth going through numpy if possible in here as well.

File: torch/distributed/distributed_c10d.py [hunk 3]
```diff
@@ -232,30 +213,6 @@ def _check_tensor_list(param, param_name):
...
-
-def is_nccl_available():
-    """
-    Checks if the NCCL backend is available.
-
-    """
-    return _NCCL_AVAILABLE
-
-
-def is_gloo_available():
```
Comment (apaszke) [inline torch/distributed/distributed_c10d.py]: Those seem useful?

File: torch/distributed/distributed_c10d.py [hunk 4]
```diff
@@ -340,9 +297,10 @@ def init_process_group(backend,
             must have exclusive access to every GPU it uses, as sharing GPUs
             between processes can result in deadlocks.
         init_method (str, optional): URL specifying how to initialize the
-                                     process group. Default is "env://" if no
-                                     ``init_method`` or ``store`` is specified.
-                                     Mutually exclusive with ``store``.
+                                     process group. Default is "env://" or
+                                     "mpi://" if no ``init_method`` or
+                                     ``store`` is specified. Mutually
```
Comment (apaszke) [inline torch/distributed/distributed_c10d.py]: What does it mean that the default is A or B?

File: torch/distributed/distributed_c10d.py [hunk 5]
```diff
@@ -379,7 +337,10 @@ def init_process_group(backend,
         assert world_size > 0, 'world_size must be positive if using store'
         assert rank >= 0, 'rank must be non-negative if using store'
     elif init_method is None:
-        init_method = "env://"
+        if is_mpi_available() and "OMPI_COMM_WORLD_SIZE" in os.environ:
```
Comment (apaszke) [inline torch/distributed/distributed_c10d.py]: This env variable seems very OpenMPI specific, while people might be using e.g. MPICH or the Intel implementations

File: torch/distributed/rendezvous.py [hunk 6]
```diff
@@ -147,6 +150,46 @@ def _env_error(var):
     raise RuntimeError("Unable to perform rerendezvous using env:// method")
 
 
+def _mpi_rendezvous_handler(url):
+    def _error(msg):
+        return _rendezvous_error("mpi:// rendezvous: " + msg)
+
+    if not url.startswith("mpi://"):
```
Comment (apaszke) [inline torch/distributed/rendezvous.py]: Shouldn't it be _equal_ to `"mpi://"`?

File: torch/distributed/rendezvous.py [hunk 7]
```diff
@@ -147,6 +150,46 @@ def _env_error(var):
...
+        raise _error("URL must start with `mpi://`")
+
+    if not is_mpi_available():
+        raise _error("MPI backend not available")
+
+    result = urlparse(url)
+    query = dict(pair.split("=") for pair in filter(None, result.query.split("&")))
+    port = int(query.get("port", 0))
+
+    def get_local_address(port=0):
```
Comment (apaszke) [inline torch/distributed/rendezvous.py]: Heh can't wait to see sysadmins of managed scientific computing environments to wonder why are there so many jobs pinging 8.8.8.8. Is there really no other way?

File: torch/distributed/_object.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+                                Must be set only on the root process.
+        root (int, optional): The rank of the root process.
+
+    Returns:
+        The broadcast object. On the root process it returns a copy of
+        the `obj` argument (after serialization and deserialization).
+
+    """
+    tensor = None
+    length = torch.empty([1], dtype=torch.uint8)
```
Comment (mrshenli) [inline torch/distributed/_object.py]: Is `uint8` large enough for the size?

**Comment by mrshenli** [comment]
closing this one as it has been too long. But there are clear value in this PR, so add it to `distributed-backlog` and will re-evaluate in quarterly planning.

**Pull request closed without merging**